cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerClient.java

```diff
@@ -0,0 +1,105 @@
+  private final String sidecarBaseURL;
+
+  @Inject
+  ArtifactLocalizerClient(CConfiguration cConf) {
+    this.sidecarBaseURL = String
+      .format("http://localhost:%d/%s/worker", cConf.getInt(Constants.ArtifactLocalizer.PORT),
+              Constants.Gateway.INTERNAL_API_VERSION_3_TOKEN);
+  }
+
+  /**
+   * Gets the location on the local filesystem for the given artifact. This method handles fetching the artifact as well
+   * as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @return The Local Location for this artifact
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getArtifactLocation(ArtifactId artifactId) throws IOException, ArtifactNotFoundException {
+    return sendRequest(artifactId, false);
+  }
+
+  /**
+   * Gets the location on the local filesystem for the directory that contains the unpacked artifact. This method
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getUnpackedArtifactLocation(ArtifactId artifactId) throws IOException, ArtifactNotFoundException {
+    return sendRequest(artifactId, true);
+  }
+
+  private File sendRequest(ArtifactId artifactId, boolean unpack) throws IOException, ArtifactNotFoundException {
+    String urlPath = String
+      .format("/artifact/namespaces/%s/artifacts/%s/versions/%s?unpack=%b", artifactId.getNamespace(),
+              artifactId.getArtifact(),
+              artifactId.getVersion(), unpack);
+    URL url = null;
+    try {
+      url = new URI(sidecarBaseURL + urlPath).toURL();
+    } catch (URISyntaxException e) {
+      throw new IOException(e);
+    }
+
+    LOG.debug("Sending request to {}", url);
+    HttpRequest httpRequest = HttpRequest.builder(HttpMethod.GET, url).build();
+    HttpResponse httpResponse = HttpRequests.execute(httpRequest);
+
+    if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+        throw new ArtifactNotFoundException(artifactId);
+      }
+      String errorMsg = httpResponse.getResponseBodyAsString();
+      throw new IOException(errorMsg);
+    }
+
+    String path = httpResponse.getResponseBodyAsString();
+    LOG.debug("ArtifactLocalizer request returned path {}", path);
+
+    return new File(path);
```

Comment:
is it possible to get empty path as a response? If so, should we handle it otherwise may get NPE

Reply:
The only way to get empty path is if there was an error on the other side but we check the response code before this section so it should be safe

Reply:
I see